### PR TITLE
Support a list of cluster names in the cluster generator

### PR DESCRIPTION
**Pull request by gebn**
This implements the functionality described in #316. Tests coming right up.

### CODE REVIEW

**Comment by gebn** [comment]
Tests implemented.

File: pkg/generators/cluster.go [hunk 1]
```diff
@@ -69,114 +65,138 @@ func (g *ClusterGenerator) GenerateParams(
...
-			secretsFound = append(secretsFound, secretForCluster)
-
-		} else if !ignoreLocalClusters {
-			// If there is no secret for the cluster, it's the local cluster, so handle it here.
-			params := map[string]string{}
-			params["name"] = cluster.Name
-			params["server"] = cluster.Server
+	// If a selector was specified, we just selected the relevant secrets, so
+	// generate params for each match.
+	if len(config.Selector.MatchExpressions) > 0 ||
```
Comment (gebn) [inline pkg/generators/cluster.go:97]: Use a `switch` instead

File: pkg/generators/cluster.go [hunk 2]
```diff
@@ -69,114 +65,138 @@ func (g *ClusterGenerator) GenerateParams(
...
+		return paramsSet, nil
+	}
 
-			for key, value := range appSetGenerator.Clusters.Values {
-				params[fmt.Sprintf("values.%s", key)] = value
+	if len(config.Names) > 0 {
+		for _, name := range config.Names {
+			secret, ok := clusterSecrets[name]
+			if !ok {
+				log.WithField("cluster", name).Warn("skipping unknown cluster name")
```
Comment (gebn) [inline pkg/generators/cluster.go:109]: Support `in-cluster`

File: api/v1alpha1/applicationset_types.go [hunk 3]
```diff
@@ -110,6 +110,10 @@ type ClusterGenerator struct {
 	Selector metav1.LabelSelector   `json:"selector,omitempty"`
 	Template ApplicationSetTemplate `json:"template,omitempty"`
 
+	// Names defines a set of cluster names to match against all clusters registered with Argo
+	// CD. Clusters today are stored as Kubernetes Secrets. The name key is used for matching.
+	Names []string `json:"names,omitempty"`
```
Comment (gebn) [inline api/v1alpha1/applicationset_types.go:115]: Ensure numeric cluster names are parsed as strings, even without quotes.

**Comment by rishabh625** [comment]
@gebn : Can u please move this PR into argocd as applicationset is [moved](https://github.com/argoproj/argo-cd/pull/8864). If u find it laborious, i can do it on your behalf with your consent

**Pull request closed without merging**